fix: relink attachments before saving doc

maharshivpatel opened the pull request:
Certain people add attachment, before filling mandatory fields which will raise Missing Fields error.

Or any other kind of errors raised by different validators due to which file is uploaded but doc is not saved.

This will lead to orphaned/mislinked files. ex. new-purchase-receipt-1

This fix changes name of new docs to `new-<doctype>-<10digithash>` after saving the document we can use this new name to find any mislinked files created in past hour and relink them to the new doc on save.

maharshivpatel committed:
fix: relink attachments before saving doc

Certain people add attachment, before filling mandatory fields
which will raise Missing Fields error.

Or any other kind of errors raised by different validators due to which
file is uploaded but doc is not saved.

This will lead to orphaned/mislinked files. ex. new-purchase-receipt-1

This fix changes name of new docs to new-<doctype>-<no>-<10digithash>
after saving the document we can use this new name to find any mislinked
files created in past hour and relink them to the new doc on save.

ankush commented on frappe/desk/form/save.py:
```diff
@@ -17,6 +17,10 @@ def savedocs(doc, action):
+	if doc.doctype not in ["DocType", "File"] and doc.name.startswith(
+		"new-" + doc.doctype.lower().replace(" ", "-")
+	):
+		doc.file_relink_temp_docname = doc.name
```
```suggestion
		doc.__temporary_name = doc.name
```

Use `_` to denote internal attribute.

ankush commented on frappe/desk/form/save.py:
```diff
@@ -17,6 +17,10 @@ def savedocs(doc, action):
 	doc = frappe.get_doc(json.loads(doc))
 	capture_doc(doc, action)
+	if doc.doctype not in ["DocType", "File"] and doc.name.startswith(
+		"new-" + doc.doctype.lower().replace(" ", "-")
```
if `doc.__islocal` is set by client we can just assume the name is temporary right? 

This checking of exact naming of temporary names feels weird.

maharshivpatel replied:
yes, just added this as safeguard.

ankush commented on frappe/public/js/frappe/model/create_new.js:
```diff
@@ -83,7 +83,8 @@ $.extend(frappe.model, {
 		var cnt = frappe.model.new_name_count;
 		if (!cnt[doctype]) cnt[doctype] = 0;
 		cnt[doctype]++;
```
Remove `cnt` if it's not being used? or is it used anywhere else?

maharshivpatel replied:
we are still using the number, {doctype}-${cnt[doctype]}${hash}

ankush commented on frappe/public/js/frappe/views/breadcrumbs.js:
```diff
@@ -186,7 +186,14 @@ frappe.breadcrumbs = {
+				.replace(/-/g, " ")
+				.replace(/\b\w/g, (l) => l.toUpperCase());
+		}
 		let form_route = `/app/${frappe.router.slug(doctype)}/${docname}`;
```
Does this routing still work if we replace the docname? 


This is useful for going back and forth while creating documents. They stay in memory iirc.

maharshivpatel replied:
yes, i tried locally and seems to be working.

ankush commented on frappe/core/doctype/file/utils.py:
```diff
@@ -357,6 +358,49 @@ def attach_files_to_document(doc: "Document", event) -> None:
+		"File",
+		{
+			"file_url": doc.get(fieldname),
+			"attached_to_name": temp_doc_name,
```
Check if temp_doc_name is truthy? if it's falsy we shouldn't attempt linking.

maharshivpatel replied:
relink_files is not called from relink_mismatched_files if temp_doc_name is false. i can also add it here if required.

ankush commented on frappe/core/doctype/file/utils.py:
```diff
@@ -357,6 +358,49 @@ def attach_files_to_document(doc: "Document", event) -> None:
+			mislinked_file,
+			field={
+				"attached_to_name": doc.name,
+				"attached_to_doctype": doc.doctype,
+				"attached_to_field": fieldname,
```
```suggestion
```

Not required?

maharshivpatel committed:
fix(minor): don't update docname

instead of updating docname use docname_title.

https://github.com/frappe/frappe/pull/22693#discussion_r1354297835

maharshivpatel committed:
fix: don't update unnecessary fields

removed attached_to_doctype and attached_to_field from set_value as those are not required.

Co-authored-by: Ankush Menat <ankushmenat@gmail.com>

maharshivpatel committed:
chore: rename __temporary_name & check temp_doc_name

* renamed file_relink_temp_docname to __temporary_name
* added check for temp_doc_name in relink_files in case someone
    calls the function it directly

maharshivpatel committed:
fix(minor): remove __temporary_name after relink

removed __temporary_name as after relink file there is no need to keep it

surajshetty3416 commented on frappe/public/js/frappe/model/create_new.js:
```diff
@@ -83,7 +83,8 @@ $.extend(frappe.model, {
 		cnt[doctype]++;
-		return frappe.router.slug(`new-${doctype}-${cnt[doctype]}`);
+		// random hash is added to idenity mislinked files when doc is not saved and file is uploaded.
+		return frappe.router.slug(`new-${doctype}-${cnt[doctype]}${frappe.utils.get_random(10)}`);
```
```suggestion
		return frappe.router.slug(`new-${doctype}-${frappe.utils.get_random(10)}`);
```
why is count required now?

maharshivpatel committed:
fix(minor): remove count from url

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

ankush commented:
@maharshivpatel add a simple test for this?

maharshivpatel committed:
fix: attach file from library

Previously, the attach from library only shared file_url, problem is we
need more information such as is_private.

fix: send name instead of file_url on file uploader and use that in
upload_file handler to get more information about the file then attach
that info to newly created file.

maharshivpatel committed:
fix: don't overwrite name for library files

name is used in many places and should not be overwritten,
so i have sent name in upload_via_file_browser as library_file_name
which can be used to get existing file.

maharshivpatel committed:
fix: cypress and add check for __islocal

changed check for __islocal as well in savedocs

As new route ends with random hash instead of numbers,
updated cypress test to just check start of data-route instead of entire path

maharshivpatel committed:
fix: cypress tests as per new hash naming

Some Cypress Tests are using hardcoded numbers in them.
As new naming is based on random hash i have changed them accordingly.

surajshetty3416 commented on cypress/integration/timeline.js:
```diff
@@ -8,7 +8,7 @@ context("Timeline", () => {
 	it("Adding new ToDo, adding new comment, verifying comment addition & deletion and deleting ToDo", () => {
 		//Adding new ToDo
-		cy.visit("/app/todo/new-todo-1");
+		cy.visit("/app/todo/new-todo");
```
Use command for such usages as we discussed?

we have
`cy.new_form`

```suggestion
		cy.new_form("ToDo");
```

maharshivpatel committed:
fix: cypress attach control library button

*   As on clear of attach field file is deleted as well
    in `Checking functionality for "Link"` test.

*   Instead of assuming that file is present from previous test.

*   create a new doc and attach a file (link) to it.
    then create another doc and check library button functionality
    and then delete both docs.

maharshivpatel committed:
fix(minor): set is_private as int instead of bool

maharshivpatel committed:
fix: cypress add test for attach before doc save

added cypress test where attachment is added
before mandatory fields are filled so doc save will fail once
and then succeed after mandatory fields are filled then check if file is
attached to correct document and no orphaned files exist for that doc.

maharshivpatel committed:
chore: use new_form command instead of visit

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

surajshetty3416 commented on cypress/integration/control_attach.js:
```diff
@@ -137,3 +163,89 @@ context("Attach Control", () => {
+	});
+	let temp_name = "";
+	let docname = "";
+	it('Checking functionality for "Attach" Where Document is not saved but File is Uploaded', () => {
```
```suggestion
	it('Attaching a file on an unsaved document', () => {
```

surajshetty3416 commented on cypress/integration/control_attach.js:
```diff
@@ -137,3 +163,89 @@ context("Attach Control", () => {
+		});
+	});
+
+	it("Check If File was uploaded correctly", () => {
```
Fix capitalisation, check other instances as well
```suggestion
	it("Check if file was uploaded correctly", () => {
```

maharshivpatel committed:
chore: typo and replace cy.visit with cy.new_form

maharshivpatel committed:
Merge branch 'fix-attach-before-doc-save' of https://github.com/maharshivpatel/frappe into fix-attach-before-doc-save

maharshivpatel committed:
Merge branch 'develop' into fix-attach-before-doc-save

surajshetty3416 commented:
Merging.

One irrelevant test is failing. Seems flaky.

surajshetty3416 commented:
@Mergifyio backport version-14-hotfix

ankush commented:
@maharshivpatel backport this manually please.

mohsinalimat commented:
@maharshivpatel What about in web-form?

maharshivpatel commented:
@mohsinalimat If i am not mistaken you are talking about this : #21543

barredterra commented:
This has subconsciously bugged me for years, thanks!
